### ISSUE TITLE
Uploading a package allowed when no key in Origin

### ISSUE THREAD

**Issue by miketheman** [habitat-sh/habitat#895]
# The Scene

An environment so secure, it requires cryptographic magic to ensure that the person that built and uploaded a package is allowed to download and run it.

# The Players

A young(-ish) man, trying really hard to create a package that he can share with the world.
A young website, trying to allow anyone to submit and download signed packages.

---

# Act I
He types in `hab pkg build` and is finally met with the pleasure of an exit code that is equal to 0.

Following directions, he then uploads the package to the website with a "swift" `hab pkg upload results/miketheman-mytutorialapp-0.1.0-20160616184336-x86_64-linux.hart`
(director's note - the results place gets a bit messy and these names are pretty _long_ and hard to tab-complete).

The package is then shuttled off to some nether regions of US-WEST-2, where some goblins will snack on it.

**~fade~**

# Act II

Zoom in on our youngish man, trying to do the Right Thing, opens up a new session somewhere, and tries to run the packaged application.

```shellsession
$ hab sup start miketheman/mytutorialapp
∵ Package for core/hab-sup not found, installing
» Installing core/hab-sup
↓ Downloading core/busybox-static/1.24.2/20160612081725
...
↓ Downloading miketheman-20160616143833 public origin key
hab-sup(ER)[src/error.rs:310:8]: 404 Not Found
```

He bursts into tears, and throws his laptop off the top of the Empire State Building, swearing off computing and moves to a farm to raise goats.

**~fade~**

# Act III

Some time later, a not-so-young man discusses his experience with others, and finds out that he wasn't alone, and that indeed, this had been identified close to launch but had not circled back to resolve.

The man's hard exterior softens, and his pent-up misery of the previous era washes out, as he finds himself surrounded by people that love him, and accept him for who he is, just a guy trying to do a thing.

All along, all that was needed was a small guard against uploading to a website when a signing key wasn't available for download.
With that addition, the wrongs are righted, and our cast of merry troubadours continues on to the next web-town to spread joy and happiness.

# _~FIN~_

**Comment by jtimberman** [comment]
Obviously we who built habitat wouldn't have run into this because we use `core` and it's always existed on the depot server.

I think it would be a real delighter as suggested by @miketheman in slack would be to detect whether the public key exists when uploading the artifact,  and if that revision isn't available, upload the one that signed the `.hart` from the studio or otherwise local machine.

**Comment by adamhjk** [comment]
Yep - @fnichol and I have talked about this before. Makes sense we should validate that the key you signed with must exist in the depot you upload to. I would do all this client side - basically hit the API to see if the key is uploaded first, if it isn't, upload it, then upload the package - otherwise just upload the package.

**Comment by metadave** [comment]
fixed via #916, will be deployed "soonish". Thanks for reporting @miketheman
